commit df8797320bed7ad9f121477e35d7e3862efd89bd
Author: John Spray <jspray@redhat.com>
Date:   Fri Oct 6 11:02:44 2017 -0400

    mgr: cut down duplication between active+standby
    
    ...by using PyModuleRunner class from ActivePyModule too.
    
    Signed-off-by: John Spray <john.spray@redhat.com>

diff --git a/src/mgr/ActivePyModule.cc b/src/mgr/ActivePyModule.cc
index 6048dc707cf..3a6fe6a9eca 100644
--- a/src/mgr/ActivePyModule.cc
+++ b/src/mgr/ActivePyModule.cc
@@ -51,60 +51,9 @@ std::string handle_pyerror()
     return extract<std::string>(formatted);
 }
 
-
-void *ServeThread::entry()
-{
-  // No need to acquire the GIL here; the module does it.
-  dout(4) << "Entering thread for " << mod->get_name() << dendl;
-  mod->serve();
-  return nullptr;
-}
-
-ActivePyModule::ActivePyModule(const std::string &module_name_,
-                               PyObject *pClass_,
-                               PyThreadState *my_ts_)
-  : module_name(module_name_),
-    pClass(pClass_),
-    pMyThreadState(my_ts_),
-    thread(this)
-{
-}
-
-ActivePyModule::~ActivePyModule()
-{
-  if (pMyThreadState.ts != nullptr) {
-    Gil gil(pMyThreadState);
-
-    Py_XDECREF(pClassInstance);
-
-    //
-    // Ideally, now, we'd be able to do this:
-    //
-    //    Py_EndInterpreter(pMyThreadState);
-    //    PyThreadState_Swap(pMainThreadState);
-    //
-    // Unfortunately, if the module has any other *python* threads active
-    // at this point, Py_EndInterpreter() will abort with:
-    //
-    //    Fatal Python error: Py_EndInterpreter: not the last thread
-    //
-    // This can happen when using CherryPy in a module, becuase CherryPy
-    // runs an extra thread as a timeout monitor, which spends most of its
-    // life inside a time.sleep(60).  Unless you are very, very lucky with
-    // the timing calling this destructor, that thread will still be stuck
-    // in a sleep, and Py_EndInterpreter() will abort.
-    //
-    // This could of course also happen with a poorly written module which
-    // made no attempt to clean up any additional threads it created.
-    //
-    // The safest thing to do is just not call Py_EndInterpreter(), and
-    // let Py_Finalize() kill everything after all modules are shut down.
-    //
-  }
-}
-
 int ActivePyModule::load(ActivePyModules *py_modules)
 {
+  assert(py_modules);
   Gil gil(pMyThreadState);
 
   // We tell the module how we name it, so that it can be consistent
@@ -129,46 +78,6 @@ int ActivePyModule::load(ActivePyModules *py_modules)
   return load_commands();
 }
 
-int ActivePyModule::serve()
-{
-  assert(pClassInstance != nullptr);
-
-  // This method is called from a separate OS thread (i.e. a thread not
-  // created by Python), so tell Gil to wrap this in a new thread state.
-  Gil gil(pMyThreadState, true);
-
-  auto pValue = PyObject_CallMethod(pClassInstance,
-      const_cast<char*>("serve"), nullptr);
-
-  int r = 0;
-  if (pValue != NULL) {
-    Py_DECREF(pValue);
-  } else {
-    derr << module_name << ".serve:" << dendl;
-    derr << handle_pyerror() << dendl;
-    return -EINVAL;
-  }
-
-  return r;
-}
-
-void ActivePyModule::shutdown()
-{
-  assert(pClassInstance != nullptr);
-
-  Gil gil(pMyThreadState);
-
-  auto pValue = PyObject_CallMethod(pClassInstance,
-      const_cast<char*>("shutdown"), nullptr);
-
-  if (pValue != NULL) {
-    Py_DECREF(pValue);
-  } else {
-    derr << "Failed to invoke shutdown() on " << module_name << dendl;
-    derr << handle_pyerror() << dendl;
-  }
-}
-
 void ActivePyModule::notify(const std::string &notify_type, const std::string &notify_id)
 {
   assert(pClassInstance != nullptr);
diff --git a/src/mgr/ActivePyModule.h b/src/mgr/ActivePyModule.h
index 839f678f752..2a3e0c8bfe0 100644
--- a/src/mgr/ActivePyModule.h
+++ b/src/mgr/ActivePyModule.h
@@ -25,6 +25,8 @@
 #include "mon/health_check.h"
 #include "mgr/Gil.h"
 
+#include "PyModuleRunner.h"
+
 #include <vector>
 #include <string>
 
@@ -43,32 +45,9 @@ public:
   ActivePyModule *handler;
 };
 
-class ServeThread : public Thread
-{
-  ActivePyModule *mod;
-
-public:
-  ServeThread(ActivePyModule *mod_)
-    : mod(mod_) {}
-
-  void *entry() override;
-};
-
-class ActivePyModule
+class ActivePyModule : public PyModuleRunner
 {
 private:
-  const std::string module_name;
-
-  // Passed in by whoever loaded our python module and looked up
-  // the symbols in it.
-  PyObject *pClass = nullptr;
-
-  // Passed in by whoever created our subinterpreter for us
-  SafeThreadState pMyThreadState = nullptr;
-
-  // Populated when we construct our instance of pClass in load()
-  PyObject *pClassInstance = nullptr;
-
   health_check_map_t health_checks;
 
   std::vector<ModuleCommand> commands;
@@ -79,16 +58,13 @@ private:
   std::string uri;
 
 public:
-  ActivePyModule(const std::string &module_name,
+  ActivePyModule(const std::string &module_name_,
       PyObject *pClass_,
-      PyThreadState *my_ts);
-  ~ActivePyModule();
-
-  ServeThread thread;
+      PyThreadState *my_ts_)
+    : PyModuleRunner(module_name_, pClass_, my_ts_)
+  {}
 
   int load(ActivePyModules *py_modules);
-  int serve();
-  void shutdown();
   void notify(const std::string &notify_type, const std::string &notify_id);
   void notify_clog(const LogEntry &le);
 
@@ -97,11 +73,6 @@ public:
     return commands;
   }
 
-  const std::string &get_name() const
-  {
-    return module_name;
-  }
-
   int handle_command(
     const cmdmap_t &cmdmap,
     std::stringstream *ds,
